pkg/sanity/controller.go

```diff
@@ -152,10 +153,213 @@ var _ = Describe("ListVolumes [Controller Server]", func() {
+			&csi.ListVolumesRequest{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(vols).NotTo(BeNil())
+		Expect(len(vols.GetEntries())).To(Equal(totalVols))
+	})
+
+	It("should return next token when a limited number of entries are requested", func() {
+		// minVolCount is the minimum number of volumes expected to exist,
+		// based on which paginated volume listing is performed.
+		minVolCount := 5
+		// maxEntried is the maximum entries in list volume request.
+		maxEntries := 2
+		// currentTotalVols is the total number of volumes at a given time. It
+		// is used to verify that all the volumes have been listed.
+		currentTotalVols := 0
+
+		// Get the number of existing volumes.
+		vols, err := c.ListVolumes(
+			context.Background(),
+			&csi.ListVolumesRequest{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(vols).NotTo(BeNil())
+
+		initialTotalVols := len(vols.GetEntries())
+		currentTotalVols = initialTotalVols
+
+		// Ensure minimum minVolCount volumes exist.
+		if initialTotalVols < minVolCount {
+
+			By("creating required new volumes")
+			requiredVols := minVolCount - initialTotalVols
+			name := "sanity"
+			for i := 1; i <= requiredVols; i++ {
+				req := &csi.CreateVolumeRequest{
+					Name: name + strconv.Itoa(i),
+					VolumeCapabilities: []*csi.VolumeCapability{
+						{
+							AccessType: &csi.VolumeCapability_Mount{
+								Mount: &csi.VolumeCapability_MountVolume{},
+							},
+							AccessMode: &csi.VolumeCapability_AccessMode{
+								Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+							},
+						},
+					},
+				}
+
+				if secrets != nil {
+					req.ControllerCreateSecrets = secrets.CreateVolumeSecret
+				}
+
+				vol, err := c.CreateVolume(context.Background(), req)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(vol).NotTo(BeNil())
+			}
+
+			// Update the current total vols count.
+			currentTotalVols += requiredVols
+		}
+
+		// Request list volumes with max entries maxEntries.
+		vols, err = c.ListVolumes(
+			context.Background(),
+			&csi.ListVolumesRequest{
+				MaxEntries: int32(maxEntries),
+			})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(vols).NotTo(BeNil())
+
+		nextToken := vols.GetNextToken()
+
+		Expect(nextToken).To(Equal(strconv.Itoa(maxEntries)))
+		Expect(len(vols.GetEntries())).To(Equal(maxEntries))
+
+		// Request list volumes with starting_token and no max entries.
+		vols, err = c.ListVolumes(
+			context.Background(),
+			&csi.ListVolumesRequest{
+				StartingToken: nextToken,
+			})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(vols).NotTo(BeNil())
+
+		// Ensure that all the remaining entries are returned at once.
+		Expect(len(vols.GetEntries())).To(Equal(currentTotalVols - maxEntries))
+
+		if initialTotalVols < minVolCount {
+
+			By("cleaning up deleting the volumes")
+			requiredVols := minVolCount - initialTotalVols
+			name := "sanity"
+			for i := 1; i <= requiredVols; i++ {
+				delReq := &csi.DeleteVolumeRequest{
+					VolumeId: name + strconv.Itoa(i),
```

Comment:
I think this is not right, `name + strconv.Itoa(i)`  is the volume name that we create, not the id, I think we can save the volume that we create in a slice, then we can use the slice info to delete the volume we create.

Reply:
Thanks!